agentendpoint/exec_task.go

```diff
@@ -0,0 +1,106 @@
+//  Copyright 2019 Google Inc. All Rights Reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+
+package agentendpoint
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+	"runtime"
+	"time"
+
+	"github.com/GoogleCloudPlatform/osconfig/config"
+
+	agentendpointpb "github.com/GoogleCloudPlatform/osconfig/_internal/gapi-cloud-osconfig-go/google.golang.org/genproto/googleapis/cloud/osconfig/agentendpoint/v1alpha1"
+)
+
+var (
+	winRoot = os.Getenv("SystemRoot")
+	sh      = "/bin/sh"
+
+	winPowershell string
+	winCmd        string
+
+	goos = runtime.GOOS
+
+	errLinuxPowerShell = errors.New("interpreter POWERSHELL cannot be used on non-Windows system")
+	errWinNoInt        = fmt.Errorf("interpreter must be specified for a Windows system")
+)
+
+type execTask struct {
+	client *Client
+
+	TaskID    string
+	Task      *execStepTask
+	StartedAt time.Time         `json:",omitempty"`
+	LogLabels map[string]string `json:",omitempty"`
+}
+
+type execStepTask struct {
+	*agentendpointpb.ExecStepTask
+}
+
+func (e *execTask) reportCompletedState(ctx context.Context, errMsg string, output *agentendpointpb.ReportTaskCompleteRequest_ExecStepTaskOutput) error {
+	req := &agentendpointpb.ReportTaskCompleteRequest{
+		TaskId:       e.TaskID,
+		TaskType:     agentendpointpb.TaskType_EXEC_STEP_TASK,
+		ErrorMessage: errMsg,
+		Output:       output,
+	}
+	if err := e.client.ReportTaskComplete(ctx, req); err != nil {
+		return fmt.Errorf("error reporting completed state: %v", err)
+	}
+	return nil
+}
+
+func (e *execTask) run(ctx context.Context) error {
+	e.StartedAt = time.Now()
+	req := &agentendpointpb.ReportTaskProgressRequest{
+		TaskId:   e.TaskID,
+		TaskType: agentendpointpb.TaskType_EXEC_STEP_TASK,
+		Progress: &agentendpointpb.ReportTaskProgressRequest_ExecStepTaskProgress{
+			ExecStepTaskProgress: &agentendpointpb.ExecStepTaskProgress{State: agentendpointpb.ExecStepTaskProgress_STARTED},
+		},
+	}
+	res, err := e.client.ReportTaskProgress(ctx, req)
+	if err != nil {
+		return fmt.Errorf("error reporting state %s: %v", agentendpointpb.ExecStepTaskProgress_STARTED, err)
+	}
+
+	if res.GetTaskDirective() == agentendpointpb.TaskDirective_STOP {
+		return e.reportCompletedState(ctx, errServerCancel.Error(), &agentendpointpb.ReportTaskCompleteRequest_ExecStepTaskOutput{
+			// TODO: Maybe there should be a canceled state instead?
```

Comment:
Seems good to have a CANCELED state. Does reporting COMPLETED here result in an overall patch job CANCEL state?

Reply:
The service handles all that, all the agent knows is that it was told to stop before the task was finished. It is most likely because of a timeout or a user cancellation but only the service knows the exact reason. My comment here is mainly to differentiate to the service that the task on this agent as canceled before completion and not actually complete